improver/calibration/rainforest_calibration.py

```diff
@@ -0,0 +1,174 @@
+
+import warnings
+
+import numpy as np
+from iris.cube import Cube, CubeList
+
+from improver import PostProcessingPlugin
+
+# Passed to choose_set_of_percentiles to set of evenly spaced percentiles
+DEFAULT_ERROR_PERCENTILES_COUNT = 19
+DEFAULT_OUTPUT_REALIZATIONS_COUNT = 100
+
+
+class ApplyRainForestsCalibration(PostProcessingPlugin):
+    """Class to calibrate input forecast given a series of RainForests tree models."""
+
+    def __init__(self, model_config_dict: dict, threads: int):
+        """Initialise the tree model variables used in the application of RainForests
+        Calibration.
+
+        Args:
+            model_config_dict:
+                Dictionary containing Rainforests model configuration variables.
+            threads:
+                Number of threads to use during prediction with tree-model objects.
+
+        Dictionary is of format::
+
+            {
+                "-50.0" : {
+                    "lightgbm_model" : "<path_to_lightgbm_model_object>",
+                    "treelite_model" : "<path_to_treelite_model_object>"
+                },
+                "-25.0" : {
+                    "lightgbm_model" : "<path_to_lightgbm_model_object>",
+                    "treelite_model" : "<path_to_treelite_model_object>"
+                },
+                ...,
+                "50.0" : {
+                    "lightgbm_model" : "<path_to_lightgbm_model_object>",
+                    "treelite_model" : "<path_to_treelite_model_object>"
+                }
+            }
+
+        The keys specify the error threshold value, while the associated values
+        are the path to the corresponding tree-model objects for that threshold.
+
+        Treelite predictors are used if treelite_runitme is an installed dependency
+        and an associated path has been provided for all thresholds, otherwise lightgbm
+        Boosters are used as the default tree model type.
+        """
+        from lightgbm import Booster
+
+        try:
+            from treelite_runtime import Predictor
+
+            self.treelite_enabled = True
+        except ModuleNotFoundError:
+            warnings.warn(
+                "Module treelite_runtime unavailable. Defaulting to using lightgbm Boosters."
+            )
+            self.treelite_enabled = False
+
+        # Dictionary keys are strings which we will use for iterating across the
+        # config dictionary, however we cast these as float to provide the error
+        # thresholds to use in processing.
+        error_thresholds = list(model_config_dict.keys())
+        self.error_thresholds = np.array(error_thresholds, dtype=np.float32)
+
+        lightgbm_model_filenames = [
+            model_config_dict[threshold].get("lightgbm_model")
+            for threshold in error_thresholds
+        ]
+        treelite_model_filenames = [
+            model_config_dict[threshold].get("treelite_model")
+            for threshold in error_thresholds
+        ]
```

Comment:
You don't need `error_thresholds` (and I don't like using the same name twice for different contents) if you use the `values()` method of the dictionary.
```suggestion
        self.error_thresholds = np.array(
            list(model_config_dict.keys()), dtype=np.float32
        )

        lightgbm_model_filenames, treelite_model_filenames = [
            [v.get(k) for v in model_config_dict.values()]
            for k in ["lightgbm_model", "treelite_model"]
        ]
```

Reply:
Good suggestion, I'll make this change. I hadn't considered this approach and it removes an unnecessary variable.

Reply:
One thing that occurred to me in mulling this over is the possibility that the user passes in a set of error thresholds that are unordered. In future steps a threshold `DimCoord` is created from `error_thresholds` which will inevitably fail on `error_thresholds` being unordered.

I'll add a function to enforce ascending order to `error_thresholds` (and the associated `tree_models` accordingly).